Source/WebCore/Modules/mediasession/MediaMetadata.cpp

```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        return -1;
+
+    if (size.maxDimension() <= m_minimumSize)
+        return 0; // Ignore images that are too small.
+
+    // We account for the aspect ratio in scoring the "best" artwork.
+    // We prefer artwork's images with a square AR.
+    double longEdge = size.maxDimension();
+    double shortEdge = size.minDimension();
+    auto aspectRatioCoefficient = shortEdge / longEdge;
+
+    if (longEdge < m_idealSize)
+        return aspectRatioCoefficient * (0.8 * (longEdge - m_minimumSize) / (m_idealSize - m_minimumSize) + 0.2);
+    return aspectRatioCoefficient * (1.0 * m_idealSize / longEdge);
+}
+
 void MediaMetadata::refreshArtworkImage()
 {
-    const auto& mediaImages = m_metadata.artwork;
-    if (mediaImages.isEmpty()) {
-        m_artworkLoader = nullptr;
+    m_artworkImageSrc = String();
+    m_artworkImage = nullptr;
+    m_artworkLoader = nullptr;
+    m_sortedSrc.clear();
+    if (m_metadata.artwork.isEmpty())
         return;
-    }
-    if (!m_session || !m_session->document() || mediaImages[0].src == m_artworkImageSrc)
+    if (!m_session || !m_session->document())
         return;
-    // FIXME: Implement a heuristic to retrieve the "best" image.
-    m_artworkImageSrc = mediaImages[0].src;
-    m_artworkLoader = makeUnique<ArtworkImageLoader>(*m_session->document(), m_artworkImageSrc, [this](Image* image) {
-        if (!image || !image->data())
+
+    // First look into the artwork's sizes attributes to attempt to determine the best score.
+    Vector<IntSize> sizes(m_metadata.artwork.size(), [&](size_t index) -> IntSize {
+        const auto& sizes = m_metadata.artwork[index].sizes;
+        if (sizes.isEmpty())
+            return { };
+        if (equalIgnoringASCIICase(sizes, "any"_s))
+            return { m_idealSize, m_idealSize }; // We prefer image tagged with "any" size.
+        IntSize size;
+        for (auto element : StringView(sizes).split(' ')) {
+            if (element.isEmpty())
+                continue;
+            auto posX = element.findIgnoringASCIICase("x"_s);
+            if (posX == notFound || !posX)
+                return { };
+            std::optional<uint32_t> width = parseInteger<uint32_t>(element.left(posX));
+            std::optional<uint32_t> height = parseInteger<uint32_t>(element.right(posX));
+            if (!width || !height)
+                return { };
+
+            IntSize newSize { int(*width), int(*height) };
+            if (size.maxDimension() < newSize.maxDimension())
+                size = newSize;
+        }
+        return size;
+    });
+
+    m_sortedSrc = { sizes.size(), [&](size_t index) -> Pair {
+        return { imageDimensionsScore(sizes[index].width(), sizes[index].height()), m_metadata.artwork[index].src };
+    } };
+
+    std::sort(m_sortedSrc.begin(), m_sortedSrc.end(), [](const Pair& a1, const Pair& a2) {
+        return a1.score > a2.score;
+    });
+
+    tryNextArtworkImage(0);
+}
+
+void MediaMetadata::tryNextArtworkImage(uint32_t index)
```

Comment:
I would tend to do the following:
```suggestion
void MediaMetadata::tryNextArtworkImage(Vector<Pair>&&, uint32_t index = 0)
```

And have the lambda capture the Vector.
Also `m_sortedSrc`and `m_artworkLoader` are useless once we finished loading. If it was captured in the lambda, it would be freed automatically by freeing `m_artworkLoader`